File: src/main/java/com/tambapps/fft4j/complex/vector/AbstractCVector.java
package com.tambapps.fft4j.complex.vector;

import com.tambapps.fft4j.complex.Complex;

import java.util.Objects;

/**
 * An abstract CVector. Useful to extends this class when creating a new CVector implementation
 */
public abstract class AbstractCVector implements CVector {

  @Override
  public boolean equals(Object o) {
    if (!(o instanceof CVector)) {
      return false;
    }
    CVector v = (CVector) o;
    if (v.getSize() != getSize()) {
      return false;
    }

    for (int i = 0; i < getSize(); i++) {
      if (!Objects.equals(getAt(i), v.getAt(i))) {
        return false;
      }
    }
    return true;
  }

  int checkedIndex(int i) {
    if (i < 0 || i >= getSize()) {
      throw new IndexOutOfBoundsException(
        String.format("Tried to access index %d of vector with size %d", i, getSize()));
    }
    return i;
  }

  @Override
  public String toString() {
    StringBuilder stringBuilder = new StringBuilder().append("[ ");
    for (int i = 0; i < getSize(); i++) {
      stringBuilder.append(getAt(i));
      if (i < getSize() - 1) {
        stringBuilder.append(", ");
      }
    }
    stringBuilder.append(" ]");
    return stringBuilder.toString();
  }

  @Override
  public CVector copy() {
    return new ArrayCVector(copiedArray());
  }

  @Override
  public CVector immutableCopy() {
    return new ImmutableCVector(copiedArray());
  }

  @Override
  public void copy(CVector dest) {
    for (int i = 0; i < getSize(); i++) {
      dest.setAt(i, getAt(i));
    }
  }

  private Complex[] copiedArray() {
    Complex[] complexes = new Complex[getSize()];
    for (int i = 0; i < complexes.length; i++) {
      complexes[i] = getAt(i);
    }
    return complexes;
  }
}

File: src/main/kotlin/com/tambapps/fft4j/complex/vector/AbstractCVector.kt
package com.tambapps.fft4j.complex.vector

import com.tambapps.fft4j.complex.Complex

/**
 * An abstract CVector. Useful to extends this class when creating a new CVector implementation
 */
abstract class AbstractCVector : CVector {

    fun checkedIndex(i: Int): Int {
        if (i < 0 || i >= size) {
            throw IndexOutOfBoundsException(
                    java.lang.String.format("Tried to access index %d of vector with size %d", i, size))
        }
        return i
    }

    override fun toString(): String {
        val stringBuilder = StringBuilder().append("[ ")
        for (i in 0 until size) {
            stringBuilder.append(get(i))
            if (i < size - 1) {
                stringBuilder.append(", ")
            }
        }
        stringBuilder.append(" ]")
        return stringBuilder.toString()
    }

    override fun copy(): CVector {
        return ArrayCVector(copiedArray())
    }

    override fun immutableCopy(): CVector {
        return ImmutableCVector(copiedArray())
    }

    override fun copy(dest: CVector) {
        for (i in 0 until size) {
            dest[i] = get(i)
        }
    }

    private fun copiedArray(): Array<Complex> {
        return Array(size) { i -> get(i)}
    }

    override fun equals(other: Any?): Boolean {
        if (other !is CVector) {
            return false
        }
        val v = other
        if (v.size != size) {
            return false
        }
        for (i in 0 until size) {
            if (get(i) != v[i]) {
                return false
            }
        }
        return true
    }
}
